File: cost.go
package acos

import (
	"context"
	"fmt"
	"strconv"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"

	"github.com/aws/aws-sdk-go-v2/service/costexplorer"
	"github.com/aws/aws-sdk-go-v2/service/costexplorer/types"
)

type CeGetCostAndUsageAPI interface {
	GetCostAndUsage(ctx context.Context, params *costexplorer.GetCostAndUsageInput, optFns ...func(*costexplorer.Options)) (*costexplorer.GetCostAndUsageOutput, error)
}

var (
	// AWS clients
	ceClient CeGetCostAndUsageAPI
)

func init() {
	ceClient = costexplorer.NewFromConfig(cfg)
}

const (
	ceDataGranularity = "DAILY"
	ceCostMetric      = "UnblendedCost"
	ceCostGroupBy     = "LINKED_ACCOUNT"
)

// AcosGetCostsOption represents options for GetCosts. The default values are:
// - ExcludeCredit : true
// - ExcludeUpfront: true
// - ExcludeRefund : false
// - ExcludeSupport: false
type AcosGetCostsOption struct {
	ExcludeCredit  bool
	ExcludeUpfront bool
	ExcludeRefund  bool
	ExcludeSupport bool

	// acos requires the following dates to show - THIS_MONTH, vs YESTERDAY, vs LAST_WEEK, and LAST_MONTH
	dates struct {
		asOf                string
		oneWeekAgo          string
		firstDayOfLastMonth string
		firstDayOfThisMonth string // Just for flagging within the sum-up logic
	}
}

func NewGetCostsOption(asOfInUTC time.Time) AcosGetCostsOption {
	opt := AcosGetCostsOption{
		ExcludeCredit:  true,
		ExcludeUpfront: true,
		ExcludeRefund:  false,
		ExcludeSupport: false,
	}

	oneWeekAgo := asOfInUTC.Add(time.Duration(-7) * 24 * time.Hour)
	year, month, _ := asOfInUTC.Date()
	firstDayOfThisMonth := time.Date(year, month, 1, 0, 0, 0, 0, asOfInUTC.Location())
	firstDayOfLastMonth := time.Date(year, month-1, 1, 0, 0, 0, 0, asOfInUTC.Location())

	dateFmt := "2006-01-02" // Use the same format as the AWS API response, "types.ResultByTime.TimePeriod.Start/End".
	opt.dates.asOf = asOfInUTC.Format(dateFmt)
	opt.dates.oneWeekAgo = oneWeekAgo.Format(dateFmt)
	opt.dates.firstDayOfThisMonth = firstDayOfThisMonth.Format(dateFmt)
	opt.dates.firstDayOfLastMonth = firstDayOfLastMonth.Format(dateFmt)
	return opt
}

// Cost represents a cost for a given account.
type Cost struct {
	AccountID                string
	AccountName              string
	LatestDailyCostIncrease  float64
	LatestWeeklyCostIncrease float64
	AmountLastMonth          float64
	AmountThisMonth          float64
}

// Costs represents a map of Cost. The map key is the account ID of the respective Cost.
type Costs map[string]Cost // map[accountId]Cost

// Group wraps up AWS Organization Group struct.
type Group types.Group

func (g *Group) getAccountId() string {
	return g.Keys[0]
}

func (g *Group) getAmount() float64 {
	if f, err := strconv.ParseFloat(*g.Metrics[ceCostMetric].Amount, 32); err == nil {
		return f
	}
	// TODO: debug log the error
	return 0
}

// GetCosts returns the costs for given accounts.
// It raises an error when the `accounts` arg doesn't contain any account.
func GetCosts(ctx context.Context, accounts Accounts, opt AcosGetCostsOption) (Costs, error) {
	accountIds := accounts.AccountIds()
	if len(accountIds) == 0 {
		return nil, fmt.Errorf("error no account to retrieve cost: GetCosts requires at least one account in Accounts")
	}
	ceOpt := acosOptToCostExplorerOpt(opt, accountIds)

	// The following GetCostAndUsage API won't return any result in some cases (e.g. when the account is newly created).
	// We create and fill the result map with the account IDs and names first, and then fill the amount later,
	// to make sure the result map always contains all the accounts.
	costs := make(map[string]Cost)
	for _, a := range accounts {
		costs[*a.Id] = Cost{
			AccountID:                *a.Id,
			AccountName:              *a.Name,
			LatestDailyCostIncrease:  0,
			LatestWeeklyCostIncrease: 0,
			AmountLastMonth:          0,
			AmountThisMonth:          0,
		}
	}

	var nextToken *string
	for {
		ceOpt.NextPageToken = nextToken

		out, err := ceClient.GetCostAndUsage(ctx, &ceOpt)
		if err != nil {
			return nil, err
		}

		thisMonth := false
		lastWeek := false
		for _, r := range out.ResultsByTime {

			if *r.TimePeriod.Start == opt.dates.firstDayOfThisMonth {
				// We assume that the AWS API returns sorted "out.ResultsByTime" slice items
				// by "r.TimePeriod.Start".
				// So we can assume that the rest of the items are also for this month, when
				// the "r.TimePeriod.Start" equals to the "first day of this month".
				//
				// See the official doc for more details about the response data structure:
				// https://docs.aws.amazon.com/aws-cost-management/latest/APIReference/API_GetCostAndUsage.html#API_GetCostAndUsage_ResponseSyntax
				thisMonth = true
			}
			// Flag "lastWeek" if the first week has passed of this month.
			if thisMonth && *r.TimePeriod.Start == opt.dates.oneWeekAgo {
				lastWeek = true
			}

			for _, g := range r.Groups {
				grp := Group(g)
				accntId := grp.getAccountId()
				amount := grp.getAmount()

				c := costs[accntId]
				if thisMonth {
					c.AmountThisMonth += amount
				} else {
					c.AmountLastMonth += amount
				}

				// Store yesterday's cost as the "latest daily cost increase".
				//
				// The types.ResultByTime item, that represents yesterday's cost, should has
				// today's date in "r.TimePeriod.End", and yesterday's date in "r.TimePeriod.Start".
				// We only check the "r.TimePeriod.End" value here, because we we called the AWS API
				// with the "DAILY" granularity.
				if *r.TimePeriod.End == opt.dates.asOf {
					if opt.dates.asOf != opt.dates.firstDayOfThisMonth { // Unless today is the first day of month.
						c.LatestDailyCostIncrease = amount
					}
				}

				// Add the cost onto the "latest weekly cost increase".
				if lastWeek {
					c.LatestWeeklyCostIncrease += amount
				}

				costs[accntId] = c
			}
		}

		nextToken = out.NextPageToken
		if nextToken == nil {
			break
		}
	}

	return costs, nil
}

// acosOptToCostExplorerOpt returns the AWS Cost Explorer's GetCostAndUsageInput param built from the acos options.
func acosOptToCostExplorerOpt(opt AcosGetCostsOption, accountIds []string) costexplorer.GetCostAndUsageInput {
	// Base input parameter
	in := costexplorer.GetCostAndUsageInput{
		Granularity: ceDataGranularity,
		Metrics:     []string{ceCostMetric},
		TimePeriod: &types.DateInterval{
			// Get the cost for the last month and this month
			Start: aws.String(opt.dates.firstDayOfLastMonth),
			End:   aws.String(opt.dates.asOf),
		},
		GroupBy: []types.GroupDefinition{
			{
				Type: types.GroupDefinitionTypeDimension,
				Key:  aws.String(ceCostGroupBy),
			},
		},
		Filter: &types.Expression{
			And: []types.Expression{
				{
					Dimensions: &types.DimensionValues{
						Key:    ceCostGroupBy,
						Values: accountIds,
					},
				},
			},
		},
	}

	// Exclude options
	v := []string{}
	if opt.ExcludeCredit {
		v = append(v, "Credit")
	}
	if opt.ExcludeUpfront {
		v = append(v, "Upfront")
	}
	if opt.ExcludeRefund {
		v = append(v, "Refund")
	}
	if opt.ExcludeSupport {
		v = append(v, "Support")
	}
	if len(v) > 0 {
		in.Filter.And = append(in.Filter.And, types.Expression{
			Not: &types.Expression{
				Dimensions: &types.DimensionValues{
					Key:    "RECORD_TYPE",
					Values: v,
				},
			},
		})
	}

	return in
}

File: src/cost.rs
use std::collections::HashMap;

use anyhow::{bail, Result};
use aws_sdk_costexplorer::types::{
    DateInterval, DimensionValues, Expression, GroupDefinition, GroupDefinitionType,
};
use aws_sdk_costexplorer::Client as CeClient;
use chrono::{Datelike, NaiveDate};
use serde::Serialize;

use crate::organization::{Accounts, AccountsExt};

const CE_DATA_GRANULARITY: aws_sdk_costexplorer::types::Granularity =
    aws_sdk_costexplorer::types::Granularity::Daily;
const CE_COST_METRIC: &str = "UnblendedCost";
const CE_COST_GROUP_BY: &str = "LINKED_ACCOUNT";

/// Options for GetCosts.
#[derive(Debug, Clone)]
pub struct GetCostsOption {
    pub exclude_credit: bool,
    pub exclude_upfront: bool,
    pub exclude_refund: bool,
    pub exclude_support: bool,

    pub as_of: String,
    pub one_week_ago: String,
    pub first_day_of_last_month: String,
    pub first_day_of_this_month: String,
}

impl GetCostsOption {
    pub fn new(as_of_date: NaiveDate) -> Self {
        let one_week_ago = as_of_date - chrono::Duration::days(7);
        let first_day_of_this_month =
            NaiveDate::from_ymd_opt(as_of_date.year(), as_of_date.month(), 1).unwrap();

        let (prev_year, prev_month) = if as_of_date.month() == 1 {
            (as_of_date.year() - 1, 12)
        } else {
            (as_of_date.year(), as_of_date.month() - 1)
        };
        let first_day_of_last_month = NaiveDate::from_ymd_opt(prev_year, prev_month, 1).unwrap();

        let fmt = "%Y-%m-%d";
        Self {
            exclude_credit: true,
            exclude_upfront: true,
            exclude_refund: false,
            exclude_support: false,
            as_of: as_of_date.format(fmt).to_string(),
            one_week_ago: one_week_ago.format(fmt).to_string(),
            first_day_of_this_month: first_day_of_this_month.format(fmt).to_string(),
            first_day_of_last_month: first_day_of_last_month.format(fmt).to_string(),
        }
    }
}

/// Represents a cost for a given account.
#[derive(Debug, Clone, Serialize)]
pub struct Cost {
    #[serde(rename = "AccountID")]
    pub account_id: String,
    #[serde(rename = "AccountName")]
    pub account_name: String,
    #[serde(rename = "LatestDailyCostIncrease")]
    pub latest_daily_cost_increase: f64,
    #[serde(rename = "LatestWeeklyCostIncrease")]
    pub latest_weekly_cost_increase: f64,
    #[serde(rename = "AmountLastMonth")]
    pub amount_last_month: f64,
    #[serde(rename = "AmountThisMonth")]
    pub amount_this_month: f64,
}

/// A map of account ID -> Cost.
pub type Costs = HashMap<String, Cost>;

/// Get costs for the given accounts.
pub async fn get_costs(
    client: &CeClient,
    accounts: &Accounts,
    opt: &GetCostsOption,
) -> Result<Costs> {
    let account_ids = accounts.account_ids();
    if account_ids.is_empty() {
        bail!("error no account to retrieve cost: GetCosts requires at least one account in Accounts");
    }

    // Initialize costs map with all accounts (some may have no billing data yet)
    let mut costs: Costs = HashMap::new();
    for (id, acc) in accounts {
        costs.insert(
            id.clone(),
            Cost {
                account_id: acc.id.clone(),
                account_name: acc.name.clone(),
                latest_daily_cost_increase: 0.0,
                latest_weekly_cost_increase: 0.0,
                amount_last_month: 0.0,
                amount_this_month: 0.0,
            },
        );
    }

    let ce_input = build_cost_explorer_input(opt, &account_ids);

    let mut next_token: Option<String> = None;
    loop {
        let mut req = client
            .get_cost_and_usage()
            .granularity(CE_DATA_GRANULARITY)
            .set_metrics(Some(vec![CE_COST_METRIC.to_string()]))
            .time_period(ce_input.time_period.clone())
            .set_group_by(Some(ce_input.group_by.clone()))
            .set_filter(Some(ce_input.filter.clone()));

        if let Some(token) = &next_token {
            req = req.next_page_token(token);
        }

        let output = req.send().await?;

        let mut this_month = false;
        let mut last_week = false;

        for r in output.results_by_time() {
            let period = r.time_period().unwrap();
            let start = period.start();
            let end = period.end();

            if start == opt.first_day_of_this_month {
                this_month = true;
            }
            if this_month && start == opt.one_week_ago {
                last_week = true;
            }

            for g in r.groups() {
                let account_id = g.keys().first().map(|s| s.as_str()).unwrap_or_default();
                let amount: f64 = g
                    .metrics()
                    .and_then(|m| m.get(CE_COST_METRIC))
                    .and_then(|v| v.amount())
                    .and_then(|a| a.parse::<f64>().ok())
                    .unwrap_or(0.0);

                if let Some(c) = costs.get_mut(account_id) {
                    if this_month {
                        c.amount_this_month += amount;
                    } else {
                        c.amount_last_month += amount;
                    }

                    // Yesterday's cost
                    if end == opt.as_of && opt.as_of != opt.first_day_of_this_month {
                        c.latest_daily_cost_increase = amount;
                    }

                    // Weekly cost increase
                    if last_week {
                        c.latest_weekly_cost_increase += amount;
                    }
                }
            }
        }

        next_token = output.next_page_token().map(|s| s.to_string());
        if next_token.is_none() {
            break;
        }
    }

    Ok(costs)
}

struct CeInput {
    time_period: DateInterval,
    group_by: Vec<GroupDefinition>,
    filter: Expression,
}

fn build_cost_explorer_input(opt: &GetCostsOption, account_ids: &[String]) -> CeInput {
    let time_period = DateInterval::builder()
        .start(&opt.first_day_of_last_month)
        .end(&opt.as_of)
        .build()
        .unwrap();

    let group_by = vec![GroupDefinition::builder()
        .r#type(GroupDefinitionType::Dimension)
        .key(CE_COST_GROUP_BY)
        .build()];

    // Build filter: account IDs AND exclude record types
    let account_filter = Expression::builder()
        .dimensions(
            DimensionValues::builder()
                .key(aws_sdk_costexplorer::types::Dimension::LinkedAccount)
                .set_values(Some(account_ids.to_vec()))
                .build(),
        )
        .build();

    let mut and_expressions = vec![account_filter];

    // Exclude record types
    let mut exclude_values = Vec::new();
    if opt.exclude_credit {
        exclude_values.push("Credit".to_string());
    }
    if opt.exclude_upfront {
        exclude_values.push("Upfront".to_string());
    }
    if opt.exclude_refund {
        exclude_values.push("Refund".to_string());
    }
    if opt.exclude_support {
        exclude_values.push("Support".to_string());
    }

    if !exclude_values.is_empty() {
        let exclude_expr = Expression::builder()
            .not(
                Expression::builder()
                    .dimensions(
                        DimensionValues::builder()
                            .key(aws_sdk_costexplorer::types::Dimension::RecordType)
                            .set_values(Some(exclude_values))
                            .build(),
                    )
                    .build(),
            )
            .build();
        and_expressions.push(exclude_expr);
    }

    let filter = Expression::builder()
        .set_and(Some(and_expressions))
        .build();

    CeInput {
        time_period,
        group_by,
        filter,
    }
}
